enhancements/baremetal/debuggability-of-baremetal-ipi-deployment.md

```diff
@@ -0,0 +1,246 @@
+  this enhancement.
+
+## Proposal
+
+Broadly, deployments fail due to problems encountered during these
+installation activities:
+
+  - Pre-bootstrap (image downloading, manifest creation, etc)
+  - Infrastructure automation (Terraform)
+  - Bootstrap
+  - Bare Metal Host Provisioning (Control Plane and Workers)
+  - Operator Deployment (i.e., those rolled out by CVO)
+
+We believe that since 4.5, pre-bootstrap errors are usually detected,
+and useful information is presented to the user about how to rectify the
+problem, so this enhancement request will focus on failures that occur
+from terraform onward.
+
+### Kinds of deployment failure
+
+#### Infrastructure Automation (Terraform)
+
+Baremetal IPI relies on terraform to provision a libvirt bootstrap
+virtual machine, and bare metal control plane hosts. We use
+terraform-provider-libvirt and terraform-provider-ironic to accomplish
+those goals.
+
+terraform-provider-ironic reports failures when it cannot reach the
+Ironic API, or a control plane host fails to provision. In both cases,
+we do not provide useful information to the user about what to do.
+
+#### Bootstrap Failures
+
+The bootstrap runs a couple of baremetal-specific services, including
+Ironic as well as a utility that populates introspection data for
+the control plane.
+
+Bootstrap typically fails for baremetal when we can't download the
+machine-os image into our local HTTP cache.  Less common, but still
+sometimes seen are that services such as dnsmasq, mariadb, ironic-api,
+ironic-conductor, or ironic-inspector fail.
+
+Failures on bootstrap services rarely result in any indication to the
+user that something went wrong other than that there was a timeout.
+
+The installer has a feature for log gathering on bootstrap failure that
+does not work on baremetal. This should be the first priority, but even
+in this case a user still needs to look into an archive containing many
+logs to identify a failure.
+
+Ideally there would be some mechanism to identify and extract useful
+information and display it to the user.
+
+#### Bare Metal Host Provisioning
+
+Whether the control plane or worker nodes, provisioning of bare metal
+hosts can fail in the same ways, although the communication path to
+provide feedback is different in each case. For the control plane,
+information about failure is presented to the user via terraform. For
+workers, it would be through information on the `BareMetalHost`
+resource, and the baremetal-operator logs.
+
+Provisioning can fail in many ways. The most difficult to troubleshoot
+are simply when we fail to hear back from a host. Buggy UEFI firmware
+may prevent PXE, a kernel could panic, or even a network cable may be
+unplugged. In these cases, we should inform the user what little
+information Ironic was able to discern, but also provide a suggestion
+that the most effective way to troubleshoot the problem is examination
+of the console of the host.
+
```

Comment:
We started experiencing a non-usual behaviour on this step (OCP 4.5 nightlies from a month ago), where Ironic already write the image on disk but the host does not get restarted. On screen you will see many glyphs that seems to be the log execution on base64. The installer get stucked there, if you let the installer continues will finish with a considerably delay (more or less 2h-2,30h), if you restart the nodes mannualy the installer will continue in a normal way...there is not feedback for the user at that point and maybe should. This is the related issue [BZ 1835423](https://bugzilla.redhat.com/show_bug.cgi?id=1835423) that seems to be already fixed.

Reply:
Thanks for the info! For this case, I think this kind of problem would be caught by addressing
https://storyboard.openstack.org/#!/story/2007664. If we rely on a callback from RHCOS after a successful boot, then we could fail earlier instead of keep waiting for things to converge when we can know the masters aren't even online.